src/Conversion/ONNXToKrnl/Math/MatMulInteger.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===--------- MatMulInteger.cpp - Lowering MatMulInteger Op --------------===//
+//
+// Copyright 2019-2023 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX MatMulInteger Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/Krnl/DialectBuilder.hpp"
+#include "src/Dialect/Krnl/KrnlHelper.hpp"
+#include "src/Dialect/Mlir/DialectBuilder.hpp"
+#include "src/Dialect/Mlir/IndexExpr.hpp"
+#include "src/Dialect/ONNX/ONNXOps/ShapeHelper.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+struct ONNXMatMulIntegerOpLowering
+    : public OpConversionPattern<ONNXMatMulIntegerOp> {
+public:
+  ONNXMatMulIntegerOpLowering(TypeConverter &typeConverter, MLIRContext *ctx)
+      : OpConversionPattern(typeConverter, ctx) {}
+
+  LogicalResult matchAndRewrite(ONNXMatMulIntegerOp mmiOp,
+      ONNXMatMulIntegerOpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const final {
+    using LocalDialectBuilder = MultiDialectBuilder<KrnlBuilder,
+        IndexExprBuilderForKrnl, MathBuilder, MemRefBuilder, OnnxBuilder>;
+    Operation *op = mmiOp.getOperation();
+    Location loc = ONNXLoc<ONNXMatMulIntegerOp>(op);
+    LocalDialectBuilder create(rewriter, loc);
+
+    ValueRange operands = adaptor.getOperands();
+    Value A = adaptor.getA();
+    Value B = adaptor.getB();
+    Value aZeroPoint = mmiOp.getAZeroPoint(); // Optional input.
+    Value bZeroPoint = mmiOp.getBZeroPoint(); // Optional input.
+
+    // Common types.
+    auto aType = A.getType().dyn_cast<ShapedType>();
+    auto aZeroPointType = aZeroPoint.getType().dyn_cast<ShapedType>();
+    auto resMemRefType = dyn_cast<MemRefType>(
+        typeConverter->convertType(mmiOp.getResult().getType()));
+    Type resElementType = resMemRefType.getElementType();
+
+    // Get shape.
+    ONNXMatMulIntegerOpShapeHelper shapeHelper(op, operands, &create.krnlIE);
+    shapeHelper.computeShapeAndAssertOnFailure();
+
+    Type i32Ty = rewriter.getIntegerType(32);
+    Type ui32Ty = rewriter.getIntegerType(32, /*isSigned=*/false);
+    assert(resElementType == i32Ty && "Output type is not i32");
+
+    Value AUInt32, BUInt32;
+    // Use i32 for onnx.Sub because onnx.Sub does not support ui32.
+    // It's because `arith` dialect does not have `subui` though it has `addui`
+    // and `mului`.
```

Comment:
as discussed in another comment thread, i32 also seems like the better choice because we want to support negative values

A and AZeroPoint can represent negative numbers after subtraction even if they are both u8 to begin with